**Hello World of TensorFlow**

Let's create our first Neural Network using TensorFlow. We will try to establish a relationship between 2 numbers. 

1. Import the necessary libraries
1. Define a simple Neural Network with 1 layer. That layer has 1 neuron. The input shape will be 1 value
1. Create some training data which has a relationship y = 5x + 3
1. Train the model based on this training data
1. Provide a new value of 'x', which the model hasn't seen before, and see how accurately the model can predict the y for this new x




**1. Import the necessary libraries**

In [1]:
import numpy as np 
import tensorflow as tf
from tensorflow import keras

print("Imported the necessary libraries...")

Imported the necessary libraries...




**2. Define and Compile the Neural Network**
* When defining the NN, we create a definition of a single layer,1 neuron NN, with 1 input value
* When compiling the NN, we need to tell our model the below 2 things:
    * LOSS : When our model is training, initially, it will try to produce some random values of 'y' for an 'x'. For the training data, we do have the actual values of 'y' available. The model will try to evaluate the predicted value of 'y' against the actual value of 'y', and measure how close were the two.
    * OPTIMIZER : The objective of this function is to provide a gentle nudge, so that the model's predicted value is closer to the actual value of 'y'. The objective is to minimize the loss function


In [2]:
# First we define the Neural Network structure. This NN a single layer,1 neuron, 1 input value
my_first_model = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])


# Next, we compile the model that we just created
my_first_model.compile(optimizer='sgd', loss='mean_squared_error')

print("Defined and compiled the model...")

Defined and compiled the model...




**3. Create some training data which has a relationship y = 5x + 3**

In [3]:
x_train = np.array([-4.0, -3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0])
y_train = np.array([-17.0, -12.0, -7.0, -2.0, 3.0, 8.0, 13.0, 18.0, 23.0, 28.0])

print("Training data created...")

Training data created...




**4. Train the model using the training values created above.** 

While we know that there is a y = 5x + 3 relationship between the values (x_train & y_train), we need to model to automatically learn this rule, without telling it explicitly.

This operation will repeat for 500 iterations (epochs). With every iteration, you will see a reduction in the loss

In [4]:
my_first_model.fit(x_train, y_train, epochs=500)

Train on 10 samples
Epoch 1/500
10/10 [==============================] - 1s 79ms/sample - loss: 256.5517
Epoch 2/500
10/10 [==============================] - 0s 255us/sample - loss: 178.3121
Epoch 3/500
10/10 [==============================] - 0s 268us/sample - loss: 124.4278
Epoch 4/500
10/10 [==============================] - 0s 254us/sample - loss: 87.2980
Epoch 5/500
10/10 [==============================] - 0s 237us/sample - loss: 61.6944
Epoch 6/500
10/10 [==============================] - 0s 240us/sample - loss: 44.0210
Epoch 7/500
10/10 [==============================] - 0s 323us/sample - loss: 31.8045
Epoch 8/500
10/10 [==============================] - 0s 270us/sample - loss: 23.3435
Epoch 9/500
10/10 [==============================] - 0s 248us/sample - loss: 17.4679
Epoch 10/500
10/10 [==============================] - 0s 306us/sample - loss: 13.3725
Epoch 11/500
10/10 [==============================] - 0s 295us/sample - loss: 10.5037
Epoch 12/500
10/10 [=====================



**5. Predict 'y' for a new 'x'**

At this point, our model has learnt the relationship that 'y' holds with 'x'. We will now provide a new value of 'x', which the model has not seen before. Let's see how well our model is able to predict 'y' for this new 'x'



In [5]:
my_first_model.predict([80])

array([[403.00064]], dtype=float32)


The new value of x which we input was 80. As per our rule (y=5x + 3), the value of y should have been
y = 5 * 80 + 3 = 403

The predicted value by our model shows us a value quite close to 403.



**And so, we have trained our first neural network model using TensorFlow!**